lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- | Assessing sizes of token bundles
+module Cardano.Wallet.Write.Tx.Balance.TokenBundleSizeAssessor
+    ( TokenBundleSizeAssessor (..)
+    , TokenBundleMaxSize (..)
+    , computeTokenBundleSerializedLengthBytes
+    , tokenBundleSizeAssessor
+    , getTokenBundleMaxSize
+    )
+    where
+
+import Prelude
+
+import Cardano.Ledger.Binary
+    ( serialize', shelleyProtVer )
+import Cardano.Wallet.Primitive.Types.Tx.Constraints
+    ( TokenBundleSizeAssessment (..)
+    , TokenBundleSizeAssessor (..)
+    , TxSize (..)
+    )
+import Cardano.Wallet.Shelley.Compatibility
+    ( toCardanoValue )
+import Cardano.Wallet.Write.Tx
+    ( PParams, RecentEra, ShelleyLedgerEra, withConstraints )
+import Control.DeepSeq
+    ( NFData )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Numeric.Natural
+    ( Natural )
+import Test.QuickCheck
+    ( Arbitrary (..), oneof )
+
+import Cardano.Ledger.Api
+    ( ppMaxValSizeL )
+import Control.Lens
+    ( (^.) )
+
+import qualified Cardano.Api.Shelley as Cardano
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Data.ByteString as BS
+
+-- | The maximum size of a serialized `TokenBundle` (`_maxValSize` in the Alonzo
+-- ledger)
+newtype TokenBundleMaxSize = TokenBundleMaxSize
+    { unTokenBundleMaxSize :: TxSize }
+    deriving (Eq, Generic, Show)
+
+instance NFData TokenBundleMaxSize
+
+instance Arbitrary TokenBundleMaxSize where
+    arbitrary = TokenBundleMaxSize . TxSize <$>
+        oneof
+          -- Generate values close to the mainnet value of 4000 (and guard
+          -- against underflow)
+          [ fromIntegral . max 0 . (4000 +) <$> arbitrary @Int
+
+          -- Generate more extreme values (both small and large)
+          , fromIntegral <$> arbitrary @Word64
+          ]
+    shrink (TokenBundleMaxSize (TxSize s)) =
+        map (TokenBundleMaxSize . TxSize . fromIntegral)
+        . shrink @Word64 -- Safe w.r.t the generator, despite TxSize wrapping a
+                         -- Natural
+        $ fromIntegral s
```

Comment:
Shouldn't this `Arbitrary` instance be defined within the `Spec` module?

(If for some reason we need to include this logic within the implementation, then I would be in favour of following our existing pattern and defining `{gen,shrink}TokenBundleMaxSize} inside a `Gen` module.)